### ISSUE TITLE
Problem with supervisor and some questions

### ISSUE THREAD

**Issue by caiokawasaki** [laravel/octane#174]
- Octane Version: ^0.1.1
- Laravel Version: ^8.12
- PHP Version: PHP 8.0.3 (cli)
- Server: Swoole
- Database Driver & Version: Not used

### Description:
First of all, I will describe my current environment so that you can have a bigger view of what happened in my tests today.

We have an API that runs on top of Laravel, we have many requests per second and we have several servers behind a Loadbalancer to be able to handle all this traffic. A long time ago we were using [LaravelS](https://github.com/hhxsv5/laravel-s), however we ended up having some memory problems with it and decided to go back to PHP-fpm.

As soon as we saw at Laracon that Octane would be released we became excited. Today in the afternoon I decided to test Octane, because it is a native Laravel package we are almost certain that the problems we had will not exist, as far as I could see the recycling of requests is working perfectly, the same did not happen with LaravelS.

My problem is basically the following: There is a specific route in my API that users call that route calls another API to send messages via WhatsApp. This secondary API is a Docker server. Sometimes this request can take from 0.5 to 10 seconds depending on the Docker server processing. I can't send this message to a queue because I need to return the message's Id to the user.

I reconstructed the flow of my API using the Octane today, it becomes something like this:

![api-schema](https://user-images.githubusercontent.com/9747004/114324607-6bc0c480-9b01-11eb-8c9b-5822c1f64a81.png)

Everything works perfectly, with only two servers behind the Load Balancer I got great results, but analyzing all this, some doubts appeared.

Inside the swoole_http.log a lot of notifications and errors appeared, I will list 3 of each below:

```
[2021-04-11 22:42:30 $954.0]    WARNING kill_timeout_process (ERRNO 9012): [Manager] Worker#122[pid=1614] exit timeout, force kill the process
[2021-04-11 22:42:30 $954.0]    WARNING kill_timeout_process (ERRNO 9012): [Manager] Worker#121[pid=1613] exit timeout, force kill the process
[2021-04-11 22:42:30 $954.0]    WARNING kill_timeout_process (ERRNO 9012): [Manager] Worker#120[pid=1612] exit timeout, force kill the process
[2021-04-11 22:42:30 $954.0]    WARNING check_worker_exit_status: worker#121[pid=1613] abnormal exit, status=0, signal=9
[2021-04-11 22:42:30 $954.0]    WARNING check_worker_exit_status: worker#122[pid=1614] abnormal exit, status=0, signal=9
[2021-04-11 22:42:30 $954.0]    WARNING check_worker_exit_status: worker#123[pid=1615] abnormal exit, status=0, signal=9
[2021-04-11 22:43:23 *2074.35]  NOTICE  end (ERRNO 1005): session[158115] is closed
[2021-04-11 22:43:23 *2075.37]  NOTICE  end (ERRNO 1005): session[158167] is closed
[2021-04-11 22:43:23 *2075.37]  NOTICE  end (ERRNO 1005): session[158172] is closed
```

What do these notices represent? I couldn't find any correct explanation for them when looking for errors on Google, or in the documentation here in the repository.

Another thing I noticed was when I put the number of Swoole workers to 120 workers, some of the workers simply never processed anything and kept their memory at a fixed value while others didn't stop working at any time. Did I configure something wrong?

Here are some hard workers:

![hard-workers](https://user-images.githubusercontent.com/9747004/114324941-90b63700-9b03-11eb-8c3e-929f2cc778f0.png)

And here are some lazzy workers:

![lazy-workers](https://user-images.githubusercontent.com/9747004/114324949-9dd32600-9b03-11eb-8964-47d53b5a941e.png)

Why 120 workers? When doing a benchmark test I realized that the cpu usage of my server that has 4 cores did not exceed 2% and the 8GB memory did not exceed 100Mb, this is due to the fact that my request has to wait for the return of the secondary API, I did some calculations and 120 processes would be smooth and there would still be resources on the machine. Correct me if I'm wrong.

The benchmark:

```ab -n 100000 -c 50 https://api.digitalocean.com/message```

The last thing that occurred to me is that when I try to restart the Supervisor process, the process keeps stopping forever and never ends and I need to restart the server. Should this behavior exist?

![supervisor](https://user-images.githubusercontent.com/9747004/114325185-1be3fc80-9b05-11eb-9185-2cfe57c2b5b1.png)

My supervisor conf:

```
[program:server-worker]
process_name=%(program_name)s_%(process_num)02d
command=php /var/www/server/artisan octane:start --workers=120 --max-requests=1000
autostart=true
autorestart=true
stopasgroup=true
killasgroup=true
user=root
numprocs=1
redirect_stderr=true
stdout_logfile=/var/www/server/storage/logs/supervisor.log
stopwaitsecs=3600
```

### Server Setup:

Basically I created a server on Digitalocean and followed the same orientation as Sail uses: https://github.com/laravel/sail/lob/1.x/runtimes/8.0/Dockerfile

Then I configured the application and Supervisor.

And my Benchmark route is as follows:

```php
Route::get('message', function () {
    $response = \Illuminate\Support\Facades\Http::post('http://10.136.220.216');

    return $response->successful() ?
        response()->json([
            'message' => 'I successfully send the message.'
        ]) :
        response()->json([
            'message' => 'I was not able to send the message.'
        ], 500);
});
```

**Comment by codingheping** [comment]
worker exit timeout, forced to terminate?
answer :Worker process exits timeout, the bottom of Swoole forcibly terminates the process.

**Comment by josephxanderson** [comment]
Can Swoole replace Supervisor? Especially given we have `Concurrent Tasks` and `Ticks / Intervals`. Maybe you don't need to use Supervisor? Just an idea.

**Comment by shoemoney** [comment]
If you are more of a hands on person I highly recommend managing this with pm2.   I have been a linux admin for almost 30 years and nightmares with supervisor....    especially cross platform.     pm2 I would bet 1000x over plus you have realtime monitoring and can take actions and if you want to fork out for pm2plus its REALLY amazing.   it even does deployment.

https://pastebin.com/b5J88zRr

Here is my config with queue worker you just put the number of procs (similar to supervisor).   Its much more reliable if you need to do remote starts and you can even set a express endpoint and restart it or monitor...

But again thats for the hands on....   

Other node solutions are nodemon and forever.   I don't know much about watchman but you could also make a system daemon in mac or on linux pretty easy for it to monitor the proccess.  Although reloading it would be a bit more difficult it would be solid.






https://pm2.io/docs/plus/overview/

**Comment by caiokawasaki** [comment]
@josephxanderson Without `Supervisor` how am I going to control `Octane` process and keep it alive?

@shoemoney I didn't know that pm2 was able to control processes that were not written in JavaScript...

**Comment by taylorotwell** [comment]
Keep in mind that any questions about how Swoole internally passes requests to workers, etc. is outside of Octane's concern and is best asked to the Swoole maintainer's themselves.
